### PR TITLE
Improvements to raft-ann-bench scripts, docs, and benchmarking implementations. 

### LINKED ISSUE #1762: [FEA] RAFT benchmarks should allow memory type to be specified at runtime
As an example, the `MemoryType` for the dataset in the CAGRA benchmarks is currently hard-coded to `HostMmap` [here](https://github.com/rapidsai/raft/issues/1762), which means anyone using the conda package to benchmark is going to end up using that (and unknowingly taking a perf hit as a result). We should allow that to be specified at runtime.

### PR DESCRIPTION
**Pull request by cjnolet**
This is just fixing merge conflicts for #1661 to continue making progress on new self-contained Python packaging. 

Closes #1762

### CODE REVIEW

**Comment by ajschmidt8** [comment]
This PR has been marked untrusted due to the changes outlined in https://github.com/rapidsai/raft/pull/1774.

It looks like the PR has unsigned commits.

You can rebase your branch to sign the unsigned commits to avoid having to use an `/ok to test` comment for every subsequent push.

The blog post below outlines how you can do this:

- https://duan.ca/2016/12/24/rebase-multiple-git-commits-with-signing/

**Review by achirkin — CHANGES_REQUESTED** [review]
Looking forward to using our new emerging benchmarks!
A few small things missed when merging on top of #1661
e.g. `cpp/bench/ann/src/common/conf.cpp` and `cpp/bench/ann/src/common/conf.h` have been removed there.

File: cpp/bench/ann/src/common/benchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  auto lib_name        = "lib" + algo + "_ann_bench.so";
+  std::string lib_path = "";
+  if (std::getenv("CONDA_PREFIX") != nullptr) {
+    auto conda_path = std::string(std::getenv("CONDA_PREFIX")) + "/bin" + "/ann/";
+    if (std::filesystem::exists(conda_path + "ANN_BENCH")) { lib_path = conda_path; }
+  }
+  if (std::getenv("RAFT_HOME") != nullptr) {
+    auto build_path = std::string(std::getenv("RAFT_HOME")) + "/cpp" + "/build/";
+    if (std::filesystem::exists(build_path + "ANN_BENCH")) { lib_path = build_path; }
+  }
```
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/bench/ann/src/common/benchmark.cpp:64]: @divyegala do you mind addressing this one? I've addressed the rest of the feedback.

File: cpp/bench/ann/src/common/conf.hpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    if (conf.contains("subset_first_row")) {
+      dataset_conf_.subset_first_row = conf.at("subset_first_row");
+    }
+    if (conf.contains("subset_size")) { dataset_conf_.subset_size = conf.at("subset_size"); }
+
+    if (conf.contains("dtype")) {
+      dataset_conf_.dtype = conf.at("dtype");
+    } else {
+      auto filename = dataset_conf_.base_file;
+      if (!filename.compare(filename.size() - 4, 24, "fbin")) {
```
Comment (achirkin) [inline cpp/bench/ann/src/common/conf.hpp:94]: a typo? (found in diff against 1661)

File: cpp/bench/ann/src/common/benchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  throw std::logic_error("Failed to find out name of the looked up function");
+}
+
+template <typename T>
+auto create_algo(const std::string& algo,
+                 const std::string& distance,
+                 int dim,
+                 const nlohmann::json& conf,
+                 const std::vector<int>& dev_list,
+                 const nlohman::json& index_conf) -> std::unique_ptr<raft::bench::ann::ANN<T>>
```
Comment (achirkin) [inline cpp/bench/ann/src/common/benchmark.cpp:85]: `const nlohman::json& index_conf` is an extra argument that wasn't added to #1661

File: cpp/bench/ann/src/common/benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+{
+  static auto fname = get_fun_name(reinterpret_cast<void*>(&create_algo<T>));
+  auto handle       = load_lib(algo);
+  auto fun_addr     = dlsym(handle, fname.c_str());
+  if (fun_addr == nullptr) {
+    throw std::runtime_error("Couldn't load the create_algo function (" + algo + ")");
+  }
+  auto fun = reinterpret_cast<decltype(&create_algo<T>)>(fun_addr);
+  return fun(algo, distance, dim, conf, dev_list, index_conf);
```
Comment (achirkin) [inline cpp/bench/ann/src/common/benchmark.cpp:94]: Same as above, remove `index_conf`

**Review by achirkin — CHANGES_REQUESTED** [review]
A couple more merge artifacts/typos.

File: cpp/bench/ann/src/common/benchmark.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/bench/ann/src/common/benchmark.cpp:64]: I think this has some room for thought. Maybe we undo the changes w.r.t. #1661 here, and add/discuss them as a follow-up PR?

[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/bench/ann/src/common/benchmark.cpp:64]: @divyegala @achirkin just to be completely clear- this only applies to the single executable code path, right?

[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/bench/ann/src/common/benchmark.cpp:64]: Yes, `common/benchmark.cpp` is the `ANN_BENCH` main file.

[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/bench/ann/src/common/benchmark.cpp:64]: So this file doesn't apply to the other `<ALGO>_ANN_BENCH` executables? If this change doesn't at all affect the default operation of the benchmarks then I"m find pulling this change out. However, if this change is needed in order for the scripts to function properly w/ the default build options in a conda environment, then I'd prefer to keep the changes in and update them later when we figure out the most ideal path forward.

[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/bench/ann/src/common/benchmark.cpp:64]: Yes, `benchmark.cpp` file is not used in `<ALGO>_ANN_BENCH` executables at all, so you're safe :)

**Review by dantegd — APPROVED** [review]
Just reviewed as well as tested the changes and things look great

File: cpp/bench/ann/src/common/benchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  throw std::logic_error("Failed to find out name of the looked up function");
+}
+
+template <typename T>
+auto create_algo(const std::string& algo,
+                 const std::string& distance,
+                 int dim,
+                 const nlohmann::json& conf,
+                 const std::vector<int>& dev_list,
+                 const nlohman::json& index_conf) -> std::unique_ptr<raft::bench::ann::ANN<T>>
```
Comment (achirkin) [inline cpp/bench/ann/src/common/benchmark.cpp:76]:
```suggestion
                 const std::vector<int>& dev_list) -> std::unique_ptr<raft::bench::ann::ANN<T>>
```

File: cpp/bench/ann/src/common/benchmark.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+{
+  static auto fname = get_fun_name(reinterpret_cast<void*>(&create_algo<T>));
+  auto handle       = load_lib(algo);
+  auto fun_addr     = dlsym(handle, fname.c_str());
+  if (fun_addr == nullptr) {
+    throw std::runtime_error("Couldn't load the create_algo function (" + algo + ")");
+  }
+  auto fun = reinterpret_cast<decltype(&create_algo<T>)>(fun_addr);
+  return fun(algo, distance, dim, conf, dev_list, index_conf);
```
Comment (achirkin) [inline cpp/bench/ann/src/common/benchmark.cpp:85]:
```suggestion
  return fun(algo, distance, dim, conf, dev_list);
```

File: cpp/bench/ann/src/common/conf.hpp (hunk 2 above)
Comment (achirkin) [inline cpp/bench/ann/src/common/conf.hpp:94]:
```suggestion
      if (!filename.compare(filename.size() - 4, 4, "fbin")) {
```

**Review by achirkin — CHANGES_REQUESTED** [review]

**Review by achirkin — APPROVED** [review]
cpp/bench changes LGTM!

**Pull request merged**